lib/pacemaker/pcmk_status.c

```diff
@@ -289,41 +224,70 @@ pcmk_status(xmlNodePtr *xml)
  * \param[in]     simple_output        Whether to use a simple output format.
  *                                     Note: This is for use by \p crm_mon only
  *                                     and is planned to be deprecated.
+ * \param[in]     timeout_ms           How long to wait for a reply from the
+ *                                     \p pacemakerd API (if 0 or negative, a
+ *                                     reasonable default will be used)
  *
  * \return Standard Pacemaker return code
  */
 int
 pcmk__status(pcmk__output_t *out, cib_t *cib,
              enum pcmk__fence_history fence_history, uint32_t show,
              uint32_t show_opts, const char *only_node, const char *only_rsc,
-             const char *neg_location_prefix, bool simple_output)
+             const char *neg_location_prefix, bool simple_output,
+             int timeout_ms)
 {
     xmlNode *current_cib = NULL;
     int rc = pcmk_rc_ok;
     stonith_t *stonith = NULL;
+    enum pcmk_pacemakerd_state state = pcmk_pacemakerd_state_invalid;
 
     if (cib == NULL) {
         return ENOTCONN;
     }
 
-    if (cib->variant == cib_native) {
-        if (cib->state == cib_connected_query || cib->state == cib_connected_command) {
-            rc = pcmk_rc_ok;
-        } else {
-            rc = pacemakerd_status(out);
+    if ((cib->variant == cib_native)
+        && (cib->state != cib_connected_query)
+        && (cib->state != cib_connected_command)) {
+
+        rc = pcmk__pacemakerd_status(out, crm_system_name,
+                                     (timeout_ms < 0)? 0 : timeout_ms,
+                                     &state);
+        switch (rc) {
+            case pcmk_rc_ok:
+                switch (state) {
+                    case pcmk_pacemakerd_state_running:
+                    case pcmk_pacemakerd_state_shutting_down:
+                        // CIB may still be available while shutting down
+                        break;
+                    default:
+                        return rc;
+                }
+                break;
+            case EREMOTEIO:
+                // @TODO: Why do we try to connect to the fencer and CIB here?
```

Comment:
I haven't figured out why `EREMOTEIO` is considered okay and we try to connect to the fencer and the CIB after getting that RC. I went ahead and preserved the existing behavior there.

Reply:
pacemakerd IPC is not proxied to remote nodes, so any attempt to contact it will always fail with EREMOTEIO when run from a remote node. However the fencer and CIB are proxied and may be available.

Reply:
Sort of, I think the idea is that if the CIB or fencer can be reached, it will immediately be cleared off the screen, but otherwise it'll be visible